core.d.ts

```diff
@@ -318,6 +318,42 @@ export type ReadableStreamWithFileType = ReadableStream & {
 	readonly fileType?: FileTypeResult;
 };
 
+/**
+Function that allows specifying custom detection mechanisms.
+
+An iterable of detectors can be provided as argument for file type detection options.
+
+The detectors are called before the default detections in the provided order.
+
+Custom detectors can be used to add new FileTypeResults or to modify return behaviour of existing FileTypeResult detections.
+
+If the detector returns `undefined`, the `tokenizer.position` should be 0 (unless it's a stream). That allows other detectors to parse the file.
```

Comment:
Not sure if I agree with the "unless it's a stream". Essentially you can iterate to other detectors if you took _a bite of the apple_. Only peek is allowed, if read you you have consumed the tokenizer, which is very similar to a stream. 

I fear this an area where we can expect a lot questions from users.

Reply:
Hey, thanks for the comment.

Yeah, guess you're right about the lot of questions.

I just mindlessly took this information from [this previous discussion](https://github.com/sindresorhus/file-type/pull/399/files#r496119460).

Let me suggest a more detailed explanation here:

If the detector returns `undefined`, the `tokenizer.position` should typically be 0. This allows easy parsing by other detectors, unless subsequent custom detectors specify otherwise. Additionally, the detector shouldn't consume the tokenizer; while peeking is non-consuming, reading is.

What do you think of this?

I'm really open to anything here!

Reply:
See also my other comment: https://github.com/sindresorhus/file-type/pull/603/files#r1356979704

I suggest something like this:

If the detector returns _no_match_ it is not allowed to read from the tokenizer (the tokeinzer.position must remain 0) otherwise following scanners will read from the wrong file offset.

If the detector return _undefined_ the scanner is certain the file type cannot be determined, neither by other scanners.

_no_match_ represents option 1 explained in [here](https://github.com/sindresorhus/file-type/pull/603/files#r1356979704)

Reply:
I agree with your point that custom detectors should be able to interrupt detection. However, I see two small downsides of the suggested approach:

1. The wording _no_match_ does not really make clear to me whether it means no match **for this detector,** or no match **at all**.

2. The standard FileTypeParser returns undefined when no file type could be recognized. Therefore requiring the custom detectors to return something else is a bit counter intuitive.

I therefore suggest to do it the other way around:

If the detector returns _undefined_, it is not allowed to read from the tokenizer (the tokenizer.position must remain 0) otherwise following scanners will read from the wrong file offset.

If the detector returns _file_type_undetectable_, the detector is certain the file type cannot be determined, even by other scanners. The FileTypeParser interrupts the parsing and immediately returns undefined.

Reply:
Okay, one could argue that _file_type_undetectable_ also does not clearly say whether it means file type undetectable **for this detector** or **for all detectors**, but it still makes it a bit clearer in my opinion.

Reply:
> I agree with your point that custom detectors should be able to interrupt detection. However, I see two small downsides of the suggested approach:
> 
> 
> 
> 1. The wording _no_match_ does not really make clear to me whether it means no match **for this detector,** or no match **at all**.
> 
> 
> 
> 2. The standard FileTypeParser returns undefined when no file type could be recognized. Therefore requiring the custom detectors to return something else is a bit counter intuitive.
> 
> 
> 
> I therefore suggest to do it the other way around:
> 
> 
> 
> If the detector returns _undefined_, it is not allowed to read from the tokenizer (the tokenizer.position must remain 0) otherwise following scanners will read from the wrong file offset.
> 
> 
> 
> If the detector returns _file_type_undetectable_, the detector is certain the file type cannot be determined, even by other scanners. The FileTypeParser interrupts the parsing and immediately returns undefined.

Sounds good to me. The second case can be also be something like, detector started reading but for some reason failed to determine the file-type. Not ideally, but it can happen. If the detector starts reading, there is no way back.

We could also check the position after each custom scanner. It may not be 0 actually, there is also an iterated use case with ID3 header. The position should be remain unchanged.

Reply:
Good idea! Just pushed a commit taking care of that check.